### PR TITLE
convert the implementation of the action to TypeScript

### PR DESCRIPTION
**Pull request by devoncarew**
Convert the implementation of the action to TypeScript:

- a mostly direct translation from the Bash implementation to TypeScript
- kept the `dart.yml` CI the same, so we could verify correct behavior of the port
- as part of the port, addressed an issue related to auto-detection of the platform architecture (fix https://github.com/dart-lang/setup-dart/issues/59)
- also, added a new action output - `dart-version`
- much of the boilerplate for the action was taken from the https://github.com/actions/typescript-action template repo
- done as part of https://github.com/dart-lang/setup-dart/issues/62
- updated the changelog w/ the relevant changes, and rev'd the action version to `1.5.0-dev`
- updated with our std CONTRIBUTING.md file from dart-lang/.github, and added a section on developing the action (building the typescript)

### CODE REVIEW

**Comment by devoncarew** [comment]
@dantup - before I ask for a general review of the implementation and such - do you mind looking over the typescript code here, and the various typescript config files? It's been a while since I've written typescript, and it would be good to have somebody w/ domain expertise look the files over :)

**Review by DanTup — APPROVED** [review]
All looks good to me. I'm not completely familiar with all the latest TS config options/files (my experience is mostly only setting up Dart-Code and keeping it running), but what's there seems reasonable (esp. if the TS action was used as a base).

Added a few questions/nits.

File: __tests__/main.test.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import * as system from '../src/system'
+import { expect, test } from '@jest/globals'
+
+test('getPlatform', () => {
+  expect(['macos', 'linux', 'windows']).toContain(system.getPlatform())
+})
+
+test('getArch', () => {
+  expect(['x64', 'arm64']).toContain(system.getArch())
```
[the remark this answers is not in the document]
Comment (devoncarew) [reply __tests__/main.test.ts]: This was taking advantage that the only likely platforms this test would run on are x64 and macos; but checking `RUNNER_ARCH` sgtm - updated.

File: src/setup.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    if (raw) {
+      core.setOutput('dart-version', sdk_utils.getVersionFromSdk(sdkPath))
+    } else {
+      core.setOutput('dart-version', version)
+    }
+
+    // Report success; print version.
+    await exec.exec('dart', ['--version'])
+  } catch (error) {
+    if (error instanceof Error) core.setFailed(error.message)
```
[the remark this answers is not in the document]
Comment (devoncarew) [reply src/setup.ts]: This was cargo culted from the template, but I agree that catching all errors, but only acting on `Error` instances, doesn't make a lot of sense. I looked at the impl for the go, node, and java setup actions; they all catch all errors here and call `setFailed` w/o any instanceof checking. I updated to match those.

File: src/setup.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+      `dartsdk-${os}-${architecture}-release.zip`
+
+    // Use a cached sdk or download and cache the sdk; using a 'raw' sdk flavor
+    // disables caching.
+    const toolName = flavor === 'raw' ? 'dart_raw' : 'dart'
+    let sdkPath = !raw ? tc.find(toolName, version, architecture) : null
+    if (sdkPath) {
+      core.info(`Using cached sdk from ${sdkPath}.`)
+    } else {
+      core.info(url)
```
[the remark this answers is not in the document]
Comment (devoncarew) [reply src/setup.ts]: Done!

**Review by devoncarew — COMMENTED** [review]
Thanks for the review! Updated w/ the noted changes.

**Comment by devoncarew** [comment]
@athomas and @mit-mit - adding you both for a general review.

@jonasfj and @isoos - you might want to look at the code around the OIDC token (mostly in the `src/oidc.ts` file).

File: src/sdk.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+import fs from 'fs'
+import path from 'path'
+
+export function getVersionFromSdk(sdkPath: string): string | null {
+  const versionFilePath = path.join(sdkPath, 'version')
+
+  if (fs.existsSync(versionFilePath)) {
+    return fs.readFileSync(versionFilePath, 'utf8').trim()
+  } else {
+    return null
```
Comment (athomas) [inline src/sdk.ts:14]: Should we throw here? Or is there a legit case where there is no `version` file?

The action will end up storing an empty string in the output in this case:
https://github.com/actions/toolkit/blob/0db3029fcfcb1393a2b3ccd14caecd7a3460af4e/packages/core/src/utils.ts#L11

File: src/setup.ts [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+import path from 'path'
+import * as core from '@actions/core'
+import * as exec from '@actions/exec'
+import * as sdk_utils from './sdk'
+import * as system from './system'
+import * as oidc from './oidc'
+import * as versions from './versions'
+import * as tc from '@actions/tool-cache'
+
+// TODO: Support version patterns of '2.19.x' to mean the latest 2.19 version.
```
Comment (athomas) [inline src/setup.ts:14]: Maybe instead we should support patterns similar to lower bounds in packages?

Comment (athomas) [inline src/setup.ts:14]: I believe mono_repo has a feature to generate tests using the lower bound of the package, maybe setup-dart could also have that feature?
version = auto -> find pubspec yaml (or perhaps lock) and use the lower bound.

File: src/versions.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+//   "version": "2.19.2",
+//   "revision": "e46b4f59490230778e907bde2eedb06b062d31be"
+// }
+
+// Query google storage for the most recent published SDK version for the given
+// channel and flavor.
+export async function latestPublishedVersion(
+  channel: string,
+  flavor: string
+): Promise<string | null> {
```
Comment (athomas) [inline src/versions.ts:19]: The `null` isn't handled at the call sites, shouldn't we throw instead?

Right now, the action will print "Installing the linux-x64 Dart SDK version null..." and then fail somewhere later (probably when it tries to download from dart-archive, which hopefully doesn't have a file at the URL containing null).

**Review by athomas — APPROVED** [review]

**Review by mit-mit — CHANGES_REQUESTED** [review]
Would like to chat a bit about this before we go ahead and make the change

**Comment by devoncarew** [comment]
Converting this PR to a draft in favor of https://github.com/dart-lang/setup-dart/pull/77.

**Pull request closed without merging**